File: sink/elasticsearch-consumer/src/database/es.writer.ts
import { Client } from "@elastic/elasticsearch";

export default class ESWriter<T> {
  private client: Client;
  private index: string;

  constructor() {
    const host = process.env.ES_HOST;
    const port = process.env.ES_PORT;
    const username = process.env.ES_USERNAME;
    const password = process.env.ES_PASSWORD;

    this.index = process.env.ES_INDEX;
    this.client = new Client({
      node: `http://${host}:${port}`,
      auth: { username, password },
    });
  }

  async updateChatMessage(eventId: string, content: T & { createdAt?: Date }) {
    try {
      await this.client.update({
        index: this.index,
        id: eventId,
        doc: content,
        doc_as_upsert: true,
      });
    } catch (error) {
      console.error("Error updating ES message:", error);
    }
  }
}

File: sink/elasticsearch-consumer/src/main/scala/database/ESWriter.scala
package consumers.es.sink
package database

import scala.concurrent.Future
// 스레드 풀 기본 지정
import scala.concurrent.ExecutionContext.Implicits.global

import com.sksamuel.elastic4s.{ElasticClient, ElasticProperties}
import com.sksamuel.elastic4s.http.JavaClient
import com.sksamuel.elastic4s.ElasticDsl.updateById
import com.sksamuel.elastic4s.Indexable
import com.sksamuel.elastic4s.handlers.update.UpdateHandlers._

import consumers.shared.config.Env.{ES_HOST, ES_PORT, ES_USERNAME, ES_PASSWORD, ES_INDEX}

class ESWriter[T](using idx: Indexable[T]) {
  private val endpoint = s"http://$ES_USERNAME:$ES_PASSWORD@$ES_HOST:$ES_PORT"
  private val props = ElasticProperties(endpoint)
  private val client = ElasticClient(JavaClient(props))

  def updateChatMessage(eventId: String, content: T): Future[?] = {
    // ES에 업데이트 요청
    client.execute(
      updateById(ES_INDEX, eventId).docAsUpsert(content)
    )
  }
}
